wasm/jit/engine.go

```diff
@@ -148,217 +401,52 @@ func (e *engine) Call(f *wasm.FunctionInstance, params ...uint64) (results []uin
-	caller := e.callFrameStack.caller
-	e.callFrameStack = caller
-
-	// If the caller is not nil, we have to go back into the caller's frame.
-	if caller != nil {
-		e.stackBasePointer = caller.stackBasePointer
-		e.stackPointer = caller.continuationStackPointer
-		e.initModuleInstance(caller.compiledFunction.source.ModuleInstance)
-	}
-}
-
-// initModuleInstance initializes the engine's state based on the given module instance.
-func (e *engine) initModuleInstance(m *wasm.ModuleInstance) {
-	if len(m.Globals) > 0 {
-		e.globalSliceAddress = uintptr(unsafe.Pointer(&m.Globals[0]))
-	}
-	if tables := m.Tables; len(tables) > 0 {
-		// WebAssembly 1.0 (MVP) has at most 1 table
-		// See https://www.w3.org/TR/wasm-core-1/#tables%E2%91%A0
-		table := tables[0]
-		if len(table.Table) > 0 {
-			e.tableSliceAddress = uintptr(unsafe.Pointer(&table.Table[0]))
-		}
-		e.tableSliceLen = uint64(len(table.Table))
-	}
-	if m.Memory != nil {
-		e.memorySliceLen = uint64(len(m.Memory.Buffer))
-		if len(m.Memory.Buffer) > 0 {
-			e.memorySliceAddress = uintptr(unsafe.Pointer(&m.Memory.Buffer[0]))
-		}
-	}
-}
-
-// jitCallStatusCode represents the result of `jitcall`.
-// This is set by the jitted native code.
-type jitCallStatusCode uint32
-
-const (
-	// jitStatusReturned means the jitcall reaches the end of function, and returns successfully.
-	jitCallStatusCodeReturned jitCallStatusCode = iota
-	// jitCallStatusCodeCallFunction means the jitcall returns to make a regular Wasm function call.
-	jitCallStatusCodeCallFunction
-	// jitCallStatusCodeCallFunction means the jitcall returns to make a builtin function call.
-	jitCallStatusCodeCallBuiltInFunction
-	// jitCallStatusCodeUnreachable means the function invocation reaches "unreachable" instruction.
-	jitCallStatusCodeUnreachable
-	// jitCallStatusCodeInvalidFloatToIntConversion means a invalid conversion of integer to floats happened.
-	jitCallStatusCodeInvalidFloatToIntConversion
-	// jitCallStatusCodeMemoryOutOfBounds means an out of bounds memory access happened.
-	jitCallStatusCodeMemoryOutOfBounds
-	// jitCallStatusCodeInvalidTableAccess means either offset to the table was out of bounds of table, or
-	// the target element in the table was uninitialized during call_indirect instruction.
-	jitCallStatusCodeInvalidTableAccess
-	// jitCallStatusCodeTypeMismatchOnIndirectCall means the type check failed during call_indirect.
-	jitCallStatusCodeTypeMismatchOnIndirectCall
-	jitCallStatusIntegerOverflow
-	jitCallStatusIntegerDivisionByZero
-)
-
-func (s jitCallStatusCode) String() (ret string) {
-	switch s {
-	case jitCallStatusCodeReturned:
-		ret = "returned"
-	case jitCallStatusCodeCallFunction:
-		ret = "call_function"
-	case jitCallStatusCodeCallBuiltInFunction:
-		ret = "call_builtin_function"
-	case jitCallStatusCodeUnreachable:
-		ret = "unreachable"
-	}
+func (e *engine) popValue() (ret uint64) {
+	ret = e.valueStack[e.valueStackContext.stackBasePointer+e.valueStackContext.stackPointer-1]
```

Comment:
this repeats, extract a function (that will end up inlined) so you can document it